lib/StripeResource.js

```diff
@@ -262,38 +262,58 @@ StripeResource.prototype = {
     return sleepSeconds * 1000;
   },
 
-  _defaultHeaders(auth, contentLength, apiVersion) {
-    let userAgentString = `Stripe/v1 NodeBindings/${this._stripe.getConstant(
-      'PACKAGE_VERSION'
-    )}`;
-
-    if (this._stripe._appInfo) {
-      userAgentString += ` ${this._stripe.getAppInfoAsString()}`;
+  _defaultIdempotencyKey(method) {
+    // If this is a POST and we allow multiple retries, ensure an idempotency key.
+    if (method === 'POST' && this._stripe.getMaxNetworkRetries() > 0) {
+      return uuid();
     }
+    return null;
+  },
 
-    const headers = {
+  _makeHeaders(
+    auth,
+    contentLength,
+    apiVersion,
+    clientUserAgent,
+    method,
+    userSuppliedHeaders
+  ) {
+    const defaultHeaders = {
       // Use specified auth token or use default from this stripe instance:
       Authorization: auth ? `Bearer ${auth}` : this._stripe.getApiField('auth'),
       Accept: 'application/json',
       'Content-Type': 'application/x-www-form-urlencoded',
       'Content-Length': contentLength,
-      'User-Agent': userAgentString,
+      'User-Agent': this._getUserAgentString(),
+      'X-Stripe-Client-User-Agent': clientUserAgent,
+      'X-Stripe-Client-Telemetry': this._getTelemetryHeader(),
+      'Stripe-Version': apiVersion,
+      'Idempotency-Key': this._defaultIdempotencyKey(method),
```

Comment:
Why don't we pass `userSuppliedheaders` into `defaultIdempotencyKey` and use whatever the user passed in (if they passed something in and after normalizing their input). It'll save us having to run `uuid()` on _every_  POST request.

Reply:
The code was uglier, and uuid() is fast:
```
> now = Date.now(); uuid(); Date.now() - now
2
> now = Date.now(); uuid(); Date.now() - now
0
```
We'd have to normalize the user headers first, and override in two places, which rubbed me the wrong way

Reply:
eg; it'd be

```js
const cleanedUserHeaders = normalize(userHeaders)
const defaultHeaders = {
//...
  'Idempotency-Key': cleanedUserHeaders['Idempotency-Key'] || this._defaultIdempotencyKey(method)
}
Object.assign(defaultHeaders, cleanedUserHeaders) // why override twice?
```

Reply:
`uuid` is fast, so agreed that this isn't a big deal 👍

Reply:
Oof! Good catch, Tino. I'll remove that in a follow-up (please let me do it 😄 I have opinions on how I want it to happen).